### PR TITLE
Improve DB Indexes and Fix Deferred Index Logging in Compact DB Mode

### PR DESCRIPTION
**Pull request by bbilger**
1. Use "primary key wihout rowid" on tiles_shallow (compact DB) which results in a clustered index which reduces the file size and improves read as well as write performance. See https://github.com/onthegomap/planetiler/issues/243
2. In deferred and compact DB mode log the correct statement to manually create the index.
3. Move the index creation into Mbtiles#createTables.
4. [enable the compact db mode by default](https://github.com/onthegomap/planetiler/commit/f01495d1aee2303f4d400999f82c4f64187dcfe6)
5. Rename the CLI switch [--defer-mbtiles-index-creation to --skip-mbtiles-index-creation](https://github.com/onthegomap/planetiler/pull/245/commits/76fa38cdddb32cce8a8ef39e284c5ad150f5ed9a) - code evolved and index creation is no longer deferred until after writing but skipped entirely. Let me know if you are not okay with this change, then I'll revert that commit.

<details>
<summary>Here are some results on experimenting with PK with and without rowid, and unique on tiles and tiles_shallow</summary>

```
NOT compact - tiles unique
0m0:06:07 INF - 	mbtiles          1m12s cpu:10m46s gc:4s avg:9
0m0:06:07 INF - 	  read    1x(9% 7s wait:1m done:3s)
0m0:06:07 INF - 	  encode 11x(68% 49s wait:14s done:3s)
0m0:06:07 INF - 	  write   1x(59% 42s sys:6s wait:24s)
0m0:06:07 INF - ----------------------------------------
0m0:06:07 INF - 	features	3.6GB
0m0:06:07 INF - 	mbtiles	3.1GB


NOT compact - tiles primary key with rowid
0m0:06:09 INF - 	mbtiles          1m11s cpu:10m44s gc:4s avg:9
0m0:06:09 INF - 	  read    1x(10% 7s wait:59s done:3s)
0m0:06:09 INF - 	  encode 11x(68% 48s wait:14s done:3s)
0m0:06:09 INF - 	  write   1x(58% 41s sys:5s wait:25s)
0m0:06:09 INF - ----------------------------------------
0m0:06:09 INF - 	features	3.6GB
0m0:06:09 INF - 	mbtiles	3.1GB


NOT compact - tiles primary key without rowid
0m0:06:12 INF - 	mbtiles          1m20s cpu:11m13s gc:4s avg:8.4
0m0:06:12 INF - 	  read    1x(9% 7s wait:1m8s done:4s)
0m0:06:12 INF - 	  encode 11x(61% 49s wait:21s done:4s)
0m0:06:12 INF - 	  write   1x(68% 55s sys:6s wait:20s)
0m0:06:12 INF - ----------------------------------------
0m0:06:12 INF - 	features	3.6GB
0m0:06:12 INF - 	mbtiles	3.3GB

compact - tiles_shallow unique
0m0:06:02 INF - 	mbtiles          1m9s cpu:10m53s gc:4s avg:9.4
0m0:06:02 INF - 	  read    1x(9% 6s wait:59s done:3s)
0m0:06:02 INF - 	  encode 11x(71% 49s wait:11s done:3s)
0m0:06:02 INF - 	  write   1x(54% 38s sys:4s wait:26s)
0m0:06:02 INF - ----------------------------------------
0m0:06:02 INF - 	features	3.6GB
0m0:06:02 INF - 	mbtiles	1.7GB


compact - tiles_shallow primary key with rowid
0m0:06:05 INF - 	mbtiles          1m9s cpu:10m48s gc:4s avg:9.3
0m0:06:05 INF - 	  read    1x(10% 7s wait:58s done:3s)
0m0:06:05 INF - 	  encode 11x(70% 49s wait:11s done:3s)
0m0:06:05 INF - 	  write   1x(54% 38s sys:4s wait:26s)
0m0:06:05 INF - ----------------------------------------
0m0:06:05 INF - 	features	3.6GB
0m0:06:05 INF - 	mbtiles	1.7GB


compact - tiles_shallow primary key without row id
0m0:06:02 INF - 	mbtiles          1m7s cpu:10m39s gc:4s avg:9.6
0m0:06:02 INF - 	  read    1x(10% 7s wait:56s done:2s)
0m0:06:02 INF - 	  encode 11x(72% 48s wait:10s done:2s)
0m0:06:02 INF - 	  write   1x(51% 34s sys:3s wait:27s)
0m0:06:02 INF - ----------------------------------------
0m0:06:02 INF - 	features	3.6GB
0m0:06:02 INF - 	mbtiles	1.3GB


BenchmarkMbtilesRead

=> "australia-compact-unique0.mbtiles" to "australia-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 49.687799812581176%
=> "australia-not-compact-unique0.mbtiles" to "australia-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 53.854178489283925%

0m0:00:12 INF - working on australia-compact-pk-without-rowid0.mbtiles
0m0:00:46 INF - readOperationsPerSecondStats: DoubleSummaryStatistics{count=10, sum=1989993.185451, min=197608.016769, average=198999.318545, max=200026.463181}
0m0:00:46 INF - working on australia-compact-pk-with-rowid0.mbtiles
0m0:01:35 INF - readOperationsPerSecondStats: DoubleSummaryStatistics{count=10, sum=1335755.139245, min=132923.920003, average=133575.513925, max=134438.289160}
0m0:01:35 INF - working on australia-compact-unique0.mbtiles
0m0:02:24 INF - readOperationsPerSecondStats: DoubleSummaryStatistics{count=10, sum=1329429.110417, min=132805.652420, average=132942.911042, max=133152.816498}
0m0:02:24 INF - working on australia-not-compact-pk-without-rowid0.mbtiles
0m0:03:13 INF - readOperationsPerSecondStats: DoubleSummaryStatistics{count=10, sum=1348371.188745, min=134128.132977, average=134837.118874, max=135565.023329}
0m0:03:13 INF - working on australia-not-compact-pk-with-rowid0.mbtiles
0m0:04:04 INF - readOperationsPerSecondStats: DoubleSummaryStatistics{count=10, sum=1290563.540387, min=128052.278996, average=129056.354039, max=129821.923011}
0m0:04:04 INF - working on australia-not-compact-unique0.mbtiles
0m0:04:54 INF - readOperationsPerSecondStats: DoubleSummaryStatistics{count=10, sum=1293428.105100, min=128693.712996, average=129342.810510, max=131027.014826}
0m0:04:54 INF - diffs
0m0:04:54 INF - "australia-not-compact-pk-with-rowid0.mbtiles" to "australia-not-compact-unique0.mbtiles": avg read operations per second improved by 0.22196231516832654%
0m0:04:54 INF - "australia-not-compact-pk-with-rowid0.mbtiles" to "australia-compact-unique0.mbtiles": avg read operations per second improved by 3.011519294762124%
0m0:04:54 INF - "australia-not-compact-pk-with-rowid0.mbtiles" to "australia-compact-pk-with-rowid0.mbtiles": avg read operations per second improved by 3.501694991726609%
0m0:04:54 INF - "australia-not-compact-pk-with-rowid0.mbtiles" to "australia-not-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 4.479256274398111%
0m0:04:54 INF - "australia-not-compact-pk-with-rowid0.mbtiles" to "australia-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 54.195676785841954%
0m0:04:54 INF - "australia-not-compact-unique0.mbtiles" to "australia-compact-unique0.mbtiles": avg read operations per second improved by 2.7833789272868756%
0m0:04:54 INF - "australia-not-compact-unique0.mbtiles" to "australia-compact-pk-with-rowid0.mbtiles": avg read operations per second improved by 3.272469028539362%
0m0:04:54 INF - "australia-not-compact-unique0.mbtiles" to "australia-not-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 4.247865299066731%
0m0:04:54 INF - "australia-not-compact-unique0.mbtiles" to "australia-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 53.854178489283925%
0m0:04:54 INF - "australia-compact-unique0.mbtiles" to "australia-compact-pk-with-rowid0.mbtiles": avg read operations per second improved by 0.47584551739487324%
0m0:04:54 INF - "australia-compact-unique0.mbtiles" to "australia-not-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 1.424828009221116%
0m0:04:54 INF - "australia-compact-unique0.mbtiles" to "australia-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 49.687799812581176%
0m0:04:54 INF - "australia-compact-pk-with-rowid0.mbtiles" to "australia-not-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 0.9444881871254864%
0m0:04:54 INF - "australia-compact-pk-with-rowid0.mbtiles" to "australia-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 48.978890440555176%
0m0:04:54 INF - "australia-not-compact-pk-without-rowid0.mbtiles" to "australia-compact-pk-without-rowid0.mbtiles": avg read operations per second improved by 47.58496785320867%



```
</details>

### CODE REVIEW

File: planetiler-benchmarks/src/main/java/com/onthegomap/planetiler/benchmarks/BenchmarkMbtilesWriter.java [hunk 1]
```diff
@@ -68,10 +68,7 @@ public static void main(String[] args) throws IOException {
       Path outputPath = getTempOutputPath();
       try (var mbtiles = Mbtiles.newWriteToFileDatabase(outputPath, config.compactDb())) {
 
-        mbtiles.createTables();
-        if (!config.deferIndexCreation()) {
-          mbtiles.addTileIndex();
-        }
+        mbtiles.createTables(config.skipIndexCreation());
```
Comment (msbarry) [inline planetiler-benchmarks/src/main/java/com/onthegomap/planetiler/benchmarks/BenchmarkMbtilesWriter.java:71]: One other thing I noticed in the mbtiles write benchmark is that it iterates through tile coordinates like this:

```java
    for (int z = 0; z <= 14; z++) {
      int maxCoord = 1 << z;
      for (int x = 0; x < maxCoord; x++) {
        for (int y = 0; y < maxCoord; y++) {
```

but the tile index order actually inverts Y - could you change to be:

```java
for (int y = maxCoord - 1; y > 0; y--) {
```

File: planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Mbtiles.java [hunk 2]
```diff
@@ -181,15 +168,14 @@ private Mbtiles execute(String... queries) {
...
-
-  public Mbtiles createTables() {
+  /**
+   * Creates the required tables (and views).
+   *
+   * @param skipIndexCreation skip index creation on some tables - in that case the indexes should be added later
+   *                          manually
+   * @return {@code this}
+   */
+  public Mbtiles createTables(boolean skipIndexCreation) {
```
Comment (msbarry) [inline planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Mbtiles.java:178]: it's a little hard to follow what the callers want with the true/false flag... could we make add a `createTablesWithoutIndex` so it's a bit more clear?

File: planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Mbtiles.java [hunk 3]
```diff
@@ -234,14 +232,39 @@ public Mbtiles createTables() {
         TILES_DATA_TABLE, TILES_SHALLOW_TABLE, TILES_SHALLOW_COL_DATA_ID, TILES_DATA_TABLE, TILES_DATA_COL_DATA_ID
       ));
     } else {
-      ddlStatements.add("create table " + TILES_TABLE + " (" + TILES_COL_Z + " integer, " + TILES_COL_X + " integer, " +
-        TILES_COL_Y + ", " + TILES_COL_DATA + " blob);");
+      // here "unique" is much more compact than a "primary key without rowid" because the tile data is part of the table
+      String tilesUniqueAddition = skipIndexCreation ? "" : """
+        , unique(%s,%s,%s)
```
Comment (msbarry) [inline planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Mbtiles.java:237]: is there a difference between `unique` and `primary key` here?

**Review by msbarry — COMMENTED** [review]
Looking good! A few minor comments...

File: planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Mbtiles.java (hunk 3 above)
Comment (bbilger) [reply planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Mbtiles.java:237]: afaik no - other than the usual stuff that you can have only one PK and cannot insert NULL
=> let me change this to primary key since it expresses its purpose better

Comment (msbarry) [reply planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Mbtiles.java]: This PR looks ready to merge - did you want to change this from `unique` to `primary key` though? I'm fine either way.

**Comment by msbarry** [comment]
Looks good, thanks for implementing this! I'll get some performance and DB size tests on the planet along with the other performance improvements before putting out the next release.
